embassy-stm32f4/src/rtc.rs

```diff
@@ -0,0 +1,290 @@
+
+pub struct RTC<T: Instance> {
+    rtc: T,
+    irq: T::Interrupt,
+    // Timestamp marks the elapsed time, excluding the current timer.
+    timestamp: u64,
+    clocks: Clocks,
+
+    /// Timestamp at which to fire alarm. u64::MAX if no alarm is scheduled.
+    alarms: Mutex<[AlarmState; ALARM_COUNT]>,
+}
+
+unsafe impl<T: Instance> Send for RTC<T> {}
+unsafe impl<T: Instance> Sync for RTC<T> {}
+
+impl<T: Instance> RTC<T> {
+    pub fn new(rtc: T, irq: T::Interrupt, clocks: Clocks) -> Self {
+        T::enable_clock();
+
+        Self {
+            rtc,
+            irq,
+            timestamp: 0,
+            clocks: clocks,
+            alarms: Mutex::new([AlarmState::new(), AlarmState::new(), AlarmState::new()]),
+        }
+    }
+
+    #[inline(always)]
+    fn reset_timestamp(&mut self) {
+        self.timestamp += self.rtc.deref().arr.read().bits() as u64;
+    }
+
+    #[inline(always)]
+    fn set_arr(&self, arr: u32) {
+        if arr != self.rtc.deref().arr.read().bits() {
+            self.rtc.deref().arr.write(|w| unsafe { w.bits(arr) });
+            self.update();
+        }
+    }
+
+    #[inline(always)]
+    fn update(&self) {
+        self.rtc.deref().cr1.modify(|_, w| w.urs().set_bit());
+        self.rtc.deref().egr.write(|w| w.ug().set_bit());
+        self.rtc.deref().cr1.modify(|_, w| w.urs().clear_bit());
+    }
+
+    pub fn start(&mut self) {
+        // stop counter
+        self.rtc.deref().cr1.modify(|_, w| w.cen().clear_bit());
+        // reset counter
+        self.rtc.deref().cnt.reset();
+
+        let frequency = 1; // timeout.into().0;
+        let pclk_mul = if T::ppre(self.clocks) == 1 { 1 } else { 2 };
+        let ticks = T::pclk(self.clocks) * pclk_mul / frequency;
+
+        let psc = ((ticks - 1) / (1 << 16)) as u16;
+        self.rtc.deref().psc.write(|w| w.psc().bits(psc));
+
+        self.set_arr(u16::MAX as u32);
+
+        // enable interrupt
+        self.rtc.deref().dier.write(|w| w.uie().set_bit());
+
+        self.irq.set_handler(
+            |ptr| unsafe {
+                let this = &mut *(ptr as *mut Self);
+                this.on_interrupt();
+            },
+            self as *const _ as *mut _,
+        );
+        self.irq.unpend();
+        self.irq.enable();
+
+        // enable "one-pulse" mode
+        // self.rtc.deref().cr1.modify(|_, w| w.opm().set_bit());
+
+        self.rtc.deref().cr1.modify(|_, w| w.cen().set_bit());
+    }
+
+    fn recompute(&self) {
+        interrupt::free(|cs| {
+            let now = self.now();
+            let mut arr = u16::MAX as u32;
+
+            for n in 0..2 {
+                let alarm = &self.alarms.borrow(cs)[n];
+                let alarm_timestamp = alarm.timestamp.get();
+
+                let diff: u64;
+                if alarm_timestamp > now {
+                    diff = alarm_timestamp - now;
+                } else {
+                    diff = 0;
+                }
+
+                if diff < 5 {
```

Comment:
Could you explain this ? it seems like you're triggering the alarm if it's less than `now` or at most 5 units bigger ?

Reply:
The alarm is triggered if it will happen in less than 5 ticks or in the past. This works based on testing, and is conservative.  If this were to be reduced, then more testing and review of the hardware manual would be needed.

Reply:
I think a better solution for the alarm would be to use a timer with the compare feature.

Reply:
> The alarm is triggered if it will happen in less than 5 ticks or in the past

This is explicitly forbidden by the Alarm trait: "When callback is called, it is **guaranteed that now() will return a value greater or equal than timestamp.**".

You're allowed to trigger the alarm later than its time, but not earlier. The reason is if you do this, the executor will "waste" cpu time spinning because the alarm is firing but `now()` says it shouldn't run tasks yet.